src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -376,19 +395,23 @@ SolutionState AddProjects(ArrayBuilder<ProjectState> projectStates)
 
             using var _1 = ArrayBuilder<ProjectId>.GetInstance(ProjectIds.Count + projectStates.Count, out var newProjectIdsBuilder);
             using var _2 = PooledHashSet<ProjectId>.GetInstance(out var addedProjectIds);
-            var newStateMapBuilder = ProjectStates.ToBuilder();
+            using var _3 = ArrayBuilder<ProjectState>.GetInstance(ProjectStates.Length + projectStates.Count, out var newProjectStatesBuilder);
```

Comment:
```suggestion
            using var _3 = ArrayBuilder<ProjectState>.GetInstance(ProjectStates.Length + projectStates.Count, out var newSortedProjectStatesBuilder);
```

Reply:
or maybe not, since you sort later.  consider calling this unsortedProjectStatedBuilder then :D